app/components/primer/alpha/modal-dialog-element.ts

```diff
@@ -0,0 +1,150 @@
+      // if #openButton is a child of a menu, we need to focus a suitable child of the menu
+      // element since it is expected for the menu to close on click
+      const menu = this.#openButton?.closest('details') || this.#openButton?.closest('action-menu')
+      if (menu) {
+        focusIfNeeded(getFocusableChild(menu))
+      } else {
+        focusIfNeeded(this.#openButton)
+      }
+      this.#openButton = undefined
+    }
+  }
+
+  get #overlayBackdrop(): HTMLElement | null {
+    if (this.parentElement?.hasAttribute('data-modal-dialog-overlay')) {
+      return this.parentElement
+    }
+
+    return null
+  }
+
+  get showButtons(): NodeList {
+    // Dialogs may also be opened from any arbitrary button with a matching show-dialog-id data attribute
+    return document.querySelectorAll(`button[data-show-dialog-id='${this.id}']`)
+  }
+
+  connectedCallback(): void {
+    if (!this.hasAttribute('role')) this.setAttribute('role', 'dialog')
+
+    const {signal} = (this.#abortController = new AbortController())
+
+    this.ownerDocument.addEventListener(
+      'click',
+      event => {
+        const target = event.target as HTMLElement
+        const clickOutsideDialog = target.closest(this.tagName) !== this
+        const button = target?.closest('button')
+        // go over this logic:
+        if (!button) {
+          if (clickOutsideDialog) {
+            // This click is outside the dialog
+            this.close()
+          }
+          return
+        }
+
+        let dialogId = button.getAttribute('data-close-dialog-id')
+        if (dialogId === this.id) {
+          this.close()
+        }
+
+        dialogId = button.getAttribute('data-submit-dialog-id')
+        if (dialogId === this.id) {
+          this.close(true)
+        }
+
+        dialogId = button.getAttribute('data-show-dialog-id')
+        if (dialogId === this.id) {
+          //TODO: see if I can remove this
+          event.stopPropagation()
+          this.#openButton = button
+          this.show()
+        }
+      },
+      {signal}
+    )
+
+    this.addEventListener('keydown', e => this.#keydown(e))
+  }
+
+  disconnectedCallback(): void {
+    this.#abortController?.abort()
+  }
+
+  show() {
+    this.open = true
+  }
+
+  close(closed = false) {
+    if (this.open === false) return
+    const eventType = closed ? 'close' : 'cancel'
+    const dialogEvent = new Event(eventType)
+    this.dispatchEvent(dialogEvent)
+    this.open = false
+  }
+
+  #keydown(event: Event) {
+    if (!(event instanceof KeyboardEvent)) return
+    if (event.isComposing) return
+
+    switch (event.key) {
+      case 'Escape':
+        if (this.open) {
+          this.close()
+          event.preventDefault()
+          event.stopPropagation()
+        }
+        break
+    }
+  }
+}
+
+declare global {
+  interface Window {
+    ModalDialogElement: typeof ModalDialogElement
+  }
+  interface HTMLElementTagNameMap {
+    'modal-dialog': ModalDialogElement
```

Comment:
_Sidenote_: I've been wondering whether we should normalize Primer custom elements with a `primer` prefix. In this case, we don't use the term "modal" to refer to dialogs except in this custom component due to the hyphen requirement. Another example is the [`tool-tip`](https://github.com/github/github/blob/master/node_modules/%40primer/view-components/app/components/primer/alpha/tool-tip-element.js) element.

Maybe `primer-dialog`, `primer-tooltip`, `primer-split-page-layout`, etc?

Related: ADR on [spelling of component names](https://github.com/github/primer/blob/main/adrs/2022-02-15-spelling-of-component-names.md)

Reply:
We also presumably need to use a different name because we'll be running the Experimental `modal-dialog` still until we've worked out how we're handling all of the bits of functionality that were added to that version. I like `primer-dialog`.